### PR TITLE
[TS Migration] Remove `EmptyObject` type

### PR DESCRIPTION
**Pull request by tienifr**
### Details

This PR investigates and removes usages of `EmptyObject` type.

### Fixed Issues

$ https://github.com/Expensify/App/issues/39124
PROPOSAL:

### Tests

1. Verify basic flows: IOU, chat, map, mention, task work properly

- [x] Verify that no errors appear in the JS console

### Offline tests

NA

### QA Steps

1. Verify basic flows: IOU, chat, map, mention, task work properly

- [x] Verify that no errors appear in the JS console

### PR Author Checklist

- [x] I linked the correct issue in the `### Fixed Issues` section above
- [x] I wrote clear testing steps that cover the changes made in this PR
    - [x] I added steps for local testing in the `Tests` section
    - [x] I added steps for the expected offline behavior in the `Offline steps` section
    - [x] I added steps for Staging and/or Production testing in the `QA steps` section
    - [x] I added steps to cover failure scenarios (i.e. verify an input displays the correct error message if the entered data is not correct)
    - [x] I turned off my network connection and tested it while offline to ensure it matches the expected behavior (i.e. verify the default avatar icon is displayed if app is offline)
    - [x] I tested this PR with a [High Traffic account](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#high-traffic-accounts) against the staging or production API to ensure there are no regressions (e.g. long loading states that impact usability).
- [x] I included screenshots or videos for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [x] I ran the tests on **all platforms** & verified they passed on:
    - [x] Android: Native
    - [x] Android: mWeb Chrome
    - [x] iOS: Native
    - [x] iOS: mWeb Safari
    - [x] MacOS: Chrome / Safari
    - [x] MacOS: Desktop
- [x] I verified there are no console errors (if there's a console error not related to the PR, report it or open an issue for it to be fixed)
- [x] I followed proper code patterns (see [Reviewing the code](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md#reviewing-the-code))
    - [x] I verified that any callback methods that were added or modified are named for what the method does and never what callback they handle (i.e. `toggleReport` and not `onIconClick`)
    - [x] I verified that the left part of a conditional rendering a React component is a boolean and NOT a string, e.g. `myBool && <MyComponent />`.
    - [x] I verified that comments were added to code that is not self explanatory
    - [x] I verified that any new or modified comments were clear, correct English, and explained "why" the code was doing something instead of only explaining "what" the code was doing.
    - [x] I verified any copy / text shown in the product is localized by adding it to `src/languages/*` files and using the [translation method](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60)
      - [x] If any non-english text was added/modified, I verified the translation was requested/reviewed in #expensify-open-source and it was approved by an internal Expensify engineer. Link to Slack message:
    - [x] I verified all numbers, amounts, dates and phone numbers shown in the product are using the [localization methods](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60-L68)
    - [x] I verified any copy / text that was added to the app is grammatically correct in English. It adheres to proper capitalization guidelines (note: only the first word of header/labels should be capitalized), and is either coming verbatim from figma or has been approved by marketing (in order to get marketing approval, ask the Bug Zero team member to add the Waiting for copy label to the issue)
    - [x] I verified proper file naming conventions were followed for any new files or renamed files. All non-platform specific files are named after what they export and are not named "index.js". All platform-specific files are named for the platform the code supports as outlined in the README.
    - [x] I verified the JSDocs style guidelines (in [`STYLE.md`](https://github.com/Expensify/App/blob/main/contributingGuides/STYLE.md#jsdocs)) were followed
- [x] If a new code pattern is added I verified it was agreed to be used by multiple Expensify engineers
- [x] I followed the guidelines as stated in the [Review Guidelines](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md)
- [x] I tested other components that can be impacted by my changes (i.e. if the PR modifies a shared library or component like `Avatar`, I verified the components using `Avatar` are working as expected)
- [x] I verified all code is DRY (the PR doesn't include any logic written more than once, with the exception of tests)
- [x] I verified any variables that can be defined as constants (ie. in CONST.js or at the top of the file that uses the constant) are defined as such
- [x] I verified that if a function's arguments changed that all usages have also been updated correctly
- [x] If any new file was added I verified that:
    - [x] The file has a description of what it does and/or why is needed at the top of the file if the code is not self explanatory
- [x] If a new CSS style is added I verified that:
    - [x] A similar style doesn't already exist
    - [x] The style can't be created with an existing [StyleUtils](https://github.com/Expensify/App/blob/main/src/styles/utils/index.ts) function (i.e. `StyleUtils.getBackgroundAndBorderStyle(theme.componentBG)`)
- [x] If the PR modifies code that runs when editing or sending messages, I tested and verified there is no unexpected behavior for all supported markdown - URLs, single line code, code blocks, quotes, headings, bold, strikethrough, and italic.
- [x] If the PR modifies a generic component, I tested and verified that those changes do not break usages of that component in the rest of the App (i.e. if a shared library or component like `Avatar` is modified, I verified that `Avatar` is working as expected in all cases)
- [x] If the PR modifies a component related to any of the existing Storybook stories, I tested and verified all stories for that component are still working as expected.
- [x] If the PR modifies a component or page that can be accessed by a direct deeplink, I verified that the code functions as expected when the deeplink is used - from a logged in and logged out account.
- [x] If the PR modifies the UI (e.g. new buttons, new UI components, changing the padding/spacing/sizing, moving components, etc) or modifies the form input styles:
    - [x] I verified that all the inputs inside a form are aligned with each other.
    - [x] I added `Design` label and/or tagged `@Expensify/design` so the design team can review the changes.
- [x] If a new page is added, I verified it's using the `ScrollView` component to make it scrollable when more elements are added to the page.
- [x] If the `main` branch was merged into this PR after a review, I tested again and verified the outcome was still expected according to the `Test` steps.

### Screenshots/Videos
<details>
<summary>Android: Native</summary>

https://github.com/Expensify/App/assets/113963320/eb95d9d8-ad3b-4292-8ebd-3b83e8140724

</details>

<details>
<summary>Android: mWeb Chrome</summary>

https://github.com/Expensify/App/assets/113963320/02be86ac-368b-44de-8580-93695a498516

</details>

<details>
<summary>iOS: Native</summary>

https://github.com/Expensify/App/assets/113963320/c464d505-5eaa-451c-9251-f0461f7ed0e9

</details>

<details>
<summary>iOS: mWeb Safari</summary>

https://github.com/Expensify/App/assets/113963320/0f015b44-482f-42ad-b74a-49eca2b67736

</details>

<details>
<summary>MacOS: Chrome / Safari</summary>

https://github.com/Expensify/App/assets/113963320/4d58c891-52c9-4b63-8efe-d6cffb2cfce9

</details>

<details>
<summary>MacOS: Desktop</summary>

https://github.com/Expensify/App/assets/113963320/de776300-923c-4b8d-9bc5-3c0761d472e8

</details>

### CODE REVIEW

**Review by blazejkustra — CHANGES_REQUESTED** [review]
Looks very promising! 🚀 

How about EmptyObject usage in `src/components/PopoverProvider/types.ts` and `src/hooks/useKeyboardShortcut.ts`?

File: src/libs/actions/IOU.ts [hunk 1]
```diff
@@ -293,7 +292,10 @@ function getReportPreviewAction(chatReportID: string, iouReportID: string): Onyx
     // Find the report preview action from the chat report
     return (
         Object.values(reportActions).find(
-            (reportAction) => reportAction && reportAction.actionName === CONST.REPORT.ACTIONS.TYPE.REPORT_PREVIEW && reportAction.originalMessage.linkedReportID === iouReportID,
+            (reportAction) =>
+                reportAction &&
+                reportAction.actionName === CONST.REPORT.ACTIONS.TYPE.REPORT_PREVIEW &&
+                (reportAction as ReportAction & OriginalMessageReportPreview).originalMessage.linkedReportID === iouReportID,
```
Comment (VickyStash) [inline src/libs/actions/IOU.ts:298]: Why do we need assertion here? I think the type is narrowed down by the check of `actionName` above
```suggestion
                reportAction.originalMessage.linkedReportID === iouReportID,
```

File: src/libs/actions/Report.ts [hunk 2]
```diff
@@ -994,12 +993,12 @@ function navigateToAndOpenReport(
     const report = isEmptyObject(chat) ? newChat : chat;
 
     // We want to pass newChat here because if anything is passed in that param (even an existing chat), we will try to create a chat on the server
-    openReport(report.reportID, '', userLogins, newChat, undefined, undefined, undefined, avatarFile);
+    openReport(report?.reportID ?? '', '', userLogins, newChat ?? {}, undefined, undefined, undefined, avatarFile);
```
Comment (VickyStash) [inline src/libs/actions/Report.ts:996]: Same question, why empty object is used here?

File: src/libs/migrations/CheckForPreviousReportActionID.ts [hunk 3]
```diff
@@ -60,6 +61,6 @@ export default function (): Promise<void | void[]> {
             onyxData[onyxKey] = {};
         });
 
-        return Onyx.multiSet(onyxData as Record<`${typeof ONYXKEYS.COLLECTION.REPORT_ACTIONS}`, Record<string, never>>);
+        return Onyx.multiSet(onyxData as Record<`${typeof ONYXKEYS.COLLECTION.REPORT_ACTIONS}`, EmptyObject>);
```
Comment (VickyStash) [inline src/libs/migrations/CheckForPreviousReportActionID.ts:64]:
```suggestion
        return Onyx.multiSet(onyxData as ReportActionsCollectionDataSet);
```

File: src/libs/OptionsListUtils.ts [hunk 4]
```diff
@@ -2050,23 +2049,23 @@ function getShareLogOptions(options: OptionList, searchValue = '', betas: Beta[]
...
-                id: personalDetail.accountID,
+                id: personalDetail?.accountID,
             },
         ],
         descriptiveText: amountText ?? '',
-        login: personalDetail.login ?? '',
-        accountID: personalDetail.accountID,
-        keyForList: String(personalDetail.accountID),
+        login: personalDetail?.login ?? '',
+        accountID: personalDetail?.accountID ?? 0,
```
Comment (VickyStash) [inline src/libs/OptionsListUtils.ts:2067]: I think we usually use -1 as a fallback value for `accountID`
```suggestion
        accountID: personalDetail?.accountID ?? -1,
```

File: src/libs/actions/Policy/Policy.ts [hunk 5]
```diff
@@ -3551,10 +3552,10 @@ function enablePolicyWorkflows(policyID: string, enabled: boolean) {
...
                     ...(!enabled
                         ? {
-                              approvalMode: policy.approvalMode,
-                              autoReporting: policy.autoReporting,
-                              harvesting: policy.harvesting,
-                              reimbursementChoice: policy.reimbursementChoice,
+                              approvalMode: policy?.approvalMode ?? null,
+                              autoReporting: policy?.autoReporting ?? null,
+                              harvesting: policy?.harvesting ?? null,
+                              reimbursementChoice: policy?.reimbursementChoice ?? null,
```
Comment (VickyStash) [inline src/libs/actions/Policy/Policy.ts:3558]:
```suggestion
                              approvalMode: policy?.approvalMode,
                              autoReporting: policy?.autoReporting,
                              harvesting: policy?.harvesting,
                              reimbursementChoice: policy?.reimbursementChoice,
```

**Comment by VickyStash** [comment]
@tienifr Could you please explain why [`receipt` typing](https://github.com/Expensify/App/pull/42851/commits/bbc5694ba41f722dc309547f729f99559f6d8d03) updates where done? Also, why do you sometimes use `OnyxEntry<Receipt>` and sometimes just `Receipt | null` during that updates?

**Comment by fabioh8010** [comment]
@tienifr bump, and we have conflicts now

**Review by mkhutornyi — COMMENTED** [review]
Conflicts again

**Comment by fabioh8010** [comment]
@tienifr kindly bump, could we try to prioritize this PR to get it moving faster? Thanks!

**Comment by blazejkustra** [comment]
@tienifr Any updates on the PR?

**Comment by blazejkustra** [comment]
@tienifr I think you need to use `undefined` instead of `null` for most cases, let me know if it works

**Comment by mountiny** [comment]
I have started a thread in here for y'all to align on what is needed https://expensify.slack.com/archives/C01GTK53T8Q/p1718383194236479

@tienifr can you have this ready on Monday?

File: src/components/AddPaymentMethodMenu.tsx [hunk 6]
```diff
@@ -65,9 +64,9 @@ function AddPaymentMethodMenu({
 
     // Users can choose to pay with business bank account in case of Expense reports or in case of P2P IOU report
     // which then starts a bottom up flow and creates a Collect workspace where the payer is an admin and payee is an employee.
-    const isIOUReport = ReportUtils.isIOUReport(iouReport ?? {});
+    const isIOUReport = ReportUtils.isIOUReport(iouReport ?? null);
```
Comment (blazejkustra) [inline src/components/AddPaymentMethodMenu.tsx:67]:
```suggestion
    const isIOUReport = ReportUtils.isIOUReport(iouReport);
```

File: src/components/AddPaymentMethodMenu.tsx [hunk 7]
```diff
@@ -65,9 +64,9 @@ function AddPaymentMethodMenu({
...
     const canUseBusinessBankAccount =
-        ReportUtils.isExpenseReport(iouReport ?? {}) || (isIOUReport && !ReportActionsUtils.hasRequestFromCurrentAccount(iouReport?.reportID ?? '-1', session?.accountID ?? -1));
+        ReportUtils.isExpenseReport(iouReport ?? null) || (isIOUReport && !ReportActionsUtils.hasRequestFromCurrentAccount(iouReport?.reportID ?? '-1', session?.accountID ?? -1));
```
Comment (blazejkustra) [inline src/components/AddPaymentMethodMenu.tsx:69]:
```suggestion
        ReportUtils.isExpenseReport(iouReport) || (isIOUReport && !ReportActionsUtils.hasRequestFromCurrentAccount(iouReport?.reportID ?? '-1', session?.accountID ?? -1));
```

File: src/libs/actions/MapboxToken.ts [hunk 8]
```diff
@@ -39,7 +39,7 @@ const setExpirationTimer = () => {
             return;
         }
         console.debug(`[MapboxToken] Fetching a new token after waiting ${REFRESH_INTERVAL / 1000 / 60} minutes`);
-        API.read(READ_COMMANDS.GET_MAPBOX_ACCESS_TOKEN, {}, {});
+        API.read(READ_COMMANDS.GET_MAPBOX_ACCESS_TOKEN, null, {});
```
Comment (blazejkustra) [inline src/libs/actions/MapboxToken.ts:42]: Let's verify that API commands work well with `null` passed here

File: src/pages/home/report/ReportActionsList.tsx [hunk 9]
```diff
@@ -604,7 +603,7 @@ function ReportActionsList({
         [isLoadingNewerReportActions, styles.chatContentScrollView, styles.chatContentScrollViewWithHeaderLoader, canShowHeader],
     );
 
-    const lastReportAction: OnyxTypes.ReportAction | EmptyObject = useMemo(() => sortedReportActions.at(-1) ?? {}, [sortedReportActions]);
+    const lastReportAction: OnyxTypes.ReportAction | null = useMemo(() => sortedReportActions.at(-1) ?? null, [sortedReportActions]);
```
Comment (blazejkustra) [inline src/pages/home/report/ReportActionsList.tsx:606]: Let's use `undefined` here instead of `null`

File: src/types/onyx/OriginalMessage.ts [hunk 10]
```diff
@@ -593,7 +593,7 @@ type OriginalMessageMergedWithCashTransaction = {
     actionName: typeof CONST.REPORT.ACTIONS.TYPE.MERGED_WITH_CASH_TRANSACTION;
 
     /** Content of the original message */
-    originalMessage: Record<string, never>; // No data is sent with this action
+    originalMessage: undefined | null; // No data is sent with this action
```
Comment (blazejkustra) [inline src/types/onyx/OriginalMessage.ts:596]: Can we use only undefined here?
```suggestion
    originalMessage: undefined; // No data is sent with this action
```

**Review by blazejkustra — CHANGES_REQUESTED** [review]
Small remarks, looks very promising!

**Review by blazejkustra — APPROVED** [review]
Looks good to me! 🚀

**Comment by blazejkustra** [comment]
@VickyStash @mkhutornyi Yours now 👍

**Review by VickyStash — APPROVED** [review]

**Comment by blazejkustra** [comment]
@mkhutornyi Please prioritize as this can get daily conflicts 🙏

**Comment by srikarparsi** [comment]
Looks like there are more conflicts already @tienifr

**Comment by srikarparsi** [comment]
@mkhutornyi please let me know after you review so that we can merge before there's conflicts again

**Comment by srikarparsi** [comment]
Looks like there's a conflict but @mkhutornyi please review/add screenshots anyway. The conflicts look relatively small and I will also review them before merging after @tienifr fixes.

**Comment by srikarparsi** [comment]
@mkhutornyi is there any update here?

**Review by mkhutornyi — APPROVED** [review]
Looks good except conflict

**Comment by srikarparsi** [comment]
Thanks @mkhutornyi, I know screenshots/videos are a little difficult for this one but if you could add some type of screenshots that would just help check off the requirement for this.

@tienifr, I'm reviewing now but if you could fix the conflict whenever you have a chance that would be great. It looks like just replacing the {} with null.

**Review by srikarparsi — DISMISSED** [review]
This looks good to me as well apart from the conflict. Thanks everyone! @tienifr please let me know once you get around to resolving the conflict and I can re-approve and merge. There seems to be a couple more now.

**Review by srikarparsi — APPROVED** [review]
Great thanks!

**Pull request merged**